File: machine/lambda.go
package machine

type Evaluator func([]Object) Object

type LazyLambda func() *Lambda

// Lambda is a wrapper that encapsulates functional environment for native
// runtime operations (e.g. arithmetic operations).
//
// Lambda implements Object, Callable.
type Lambda struct {
	argc int
	argi int
	args []Object
	eval Evaluator
}

func NewLambda(argc int, eval Evaluator) *Lambda {
	return &Lambda{
		argc: argc,
		argi: 0,
		args: make([]Object, argc),
		eval: eval,
	}
}

func (c *Lambda) Object() {}

func (c *Lambda) Apply(arg Object) {
	c.args[c.argi] = arg
	c.argi++
}

func (c *Lambda) Call() Object {
	return c.eval(c.args)
}

File: engine/src/lambda.rs
use crate::object::{Callable, Object};

pub struct Lambda {
    args: Vec<Object>,
    eval: Eval,
}

pub type Eval = fn(&Vec<Object>) -> Object;

impl Callable for Lambda {
    fn apply(&mut self, o: Object) {
        self.args.push(o);
    }

    fn call(&mut self) -> Object {
        (self.eval)(&self.args)
    }
}

impl Lambda {
    pub fn new(eval: Eval) -> Self {
        Self {
            args: Vec::new(),
            eval: eval,
        }
    }
}
